src/mesh/domain.hpp

```diff
@@ -0,0 +1,181 @@
+
+      IndexShape(const int & nx1, const int & nx2, const int & ng) 
+      : IndexShape( std::vector<int> {nx1,nx2}, ng) {};
+
+      IndexShape(const int & nx1, const int & ng) 
+      : IndexShape( std::vector<int> {nx1}, ng) {};
+
+      IndexShape(const std::vector<int> & interior_dims, const int & ng) { 
+        assert(interior_dims.size()<=NDIM && "IndexShape cannot be initialized, the number of "
+              "dimensions exceeds the statically set dimensions, you will need to change the NDIM "
+              "constant.");
+        for( int dim=1, index=0; dim<=NDIM; ++dim, ++index){
+          if (dim <= interior_dims.size()) {
+            assert( interior_dims.at(index) > 0 && "IndexShape cannot be initialized with fewer "
+                "than 1 interior cells for each dimension");
+            x_[index].start = ng;
+            x_[index].end = x_[index].start + interior_dims.at(index) - 1;
+            entire_ncells_[index] = interior_dims.at(index) + 2*ng;
+          } else {
+            x_[index].start = 0;
+            x_[index].end = 0;
+            entire_ncells_[index] = 1;
+          }
+        }
+      };
+
+      std::array<IndexRange,NDIM> GetBounds(const IndexShapeType & type) const noexcept {
+        if( type==IndexShapeType::entire ){
+          std::array<IndexRange,NDIM> bounds;
+          for (int index=0; index<NDIM;  ++index ){
+            bounds[index].start = 0;
+            bounds[index].end = entire_ncells_[index];
+          }
+          return bounds;
+        }
+        return x_; 
+      }
+      
+      inline int is(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? 0 : x_[0].start; }
+      
+      inline int js(const IndexShapeType & type) const
+      { return (type==IndexShapeType::entire) ? 0 : x_[1].start; }
+      
+      inline int ks(const IndexShapeType & type) const
+      { return (type==IndexShapeType::entire) ? 0 : x_[2].start; }
+      
+      inline int ie(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[0]-1 : x_[0].end; }
+      
+      inline int je(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[1]-1 : x_[1].end; }
+      
+      inline int ke(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[2]-1 : x_[2].end; }
+
+      inline int nx1(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[0] : x_[0].ncells(); }
+      
+      inline int nx2(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[1] : x_[1].ncells(); }
+      
+      inline int nx3(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[2] : x_[2].ncells(); }
+
+
+      void GetIndices(const IndexShapeType & type, 
+          int & is, int & ie, int & js, int & je, int & ks, int & ke) const {
```

Comment:
I don't like non-const references and I think they violate the Google C++ style guide.

Reply:
What is the alternative? When a function is many valued, there aren't great options. Returning an array or something similar?

Reply:
I think what he means, is that it is clearer if one reserves pointers for passing in variables that will be changed and leaving const references to be passed in when the variable will not change. This follows the google style guide.